File: Harvest/Utilities/ScreenshotEncoder.swift
import UIKit

/// Turns a picked image into something cheap enough to send to a vision model.
///
/// Phone screenshots are far larger than the model needs, and vision cost
/// scales with resolution, so everything is downscaled before encoding. The
/// result is a `data:` URL — the image is never written to disk or uploaded to
/// storage.
enum ScreenshotEncoder {
    /// Longest side, in pixels, after downscaling. Matches `MAX_DIMENSION` in
    /// ScreenshotEncoder.kt: chat text stays legible well below a full-res
    /// screenshot, and both platforms must send comparable detail.
    static let maxDimension: CGFloat = 1400
    static let jpegQuality: CGFloat = 0.7
    /// Refuse anything still this large once encoded; the request would be
    /// rejected upstream anyway, and failing here gives a better message.
    static let maxEncodedBytes = 4 * 1024 * 1024

    enum EncodingError: LocalizedError {
        case couldNotEncode
        case tooLarge

        var errorDescription: String? {
            switch self {
            case .couldNotEncode:
                return "That image couldn't be read. Try a different screenshot."
            case .tooLarge:
                return "That image is too large. Try a smaller screenshot."
            }
        }
    }

    /// Longest-edge target for a send of `imageCount` images. Text in a phone
    /// screenshot stays legible at 900px, which is what makes the top of the
    /// ladder affordable. Mirrors `targetDimension` in ScreenshotEncoder.kt —
    /// a count of zero or less falls in the first rung, as it does there.
    static func targetDimension(imageCount: Int) -> CGFloat {
        if imageCount <= 2 { return maxDimension }
        if imageCount <= 5 { return 1100 }
        return 900
    }

    /// A `data:image/jpeg;base64,...` URL suitable for an `image_url` part.
    static func dataURL(from image: UIImage, target: CGFloat = maxDimension) throws -> String {
        let scaled = downscaled(image, limit: target)
        guard let jpeg = scaled.jpegData(compressionQuality: jpegQuality) else {
            throw EncodingError.couldNotEncode
        }
        guard jpeg.count <= maxEncodedBytes else {
            throw EncodingError.tooLarge
        }
        return "data:image/jpeg;base64," + jpeg.base64EncodedString()
    }

    /// Aspect-preserving downscale so the longest side is at most
    /// `maxDimension`. Images already smaller are returned untouched.
    static func downscaled(_ image: UIImage, limit: CGFloat = maxDimension) -> UIImage {
        let size = image.size
        let longest = max(size.width, size.height)
        guard longest > limit, longest > 0 else { return image }

        let ratio = limit / longest
        let target = CGSize(width: (size.width * ratio).rounded(),
                            height: (size.height * ratio).rounded())

        let format = UIGraphicsImageRendererFormat.default()
        format.scale = 1   // target is already in pixels
        return UIGraphicsImageRenderer(size: target, format: format).image { _ in
            image.draw(in: CGRect(origin: .zero, size: target))
        }
    }
}

File: android/app/src/main/java/com/harvestglass/harvest/util/ScreenshotEncoder.kt
package com.harvestglass.harvest.util

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.util.Base64
import java.io.ByteArrayOutputStream
import kotlin.math.max
import kotlin.math.roundToInt

/**
 * Turns a picked image into the inline `data:` URL the Gardener sends to
 * OpenAI. Android counterpart to ScreenshotEncoder.swift.
 *
 * Nothing touches disk and nothing is uploaded to storage — the bytes live in
 * memory for one request and are dropped.
 */
object ScreenshotEncoder {

    /** Long edge cap. Chat text stays legible well below a full-res screenshot. */
    const val MAX_DIMENSION = 1400

    /** JPEG quality; enough for text, small enough to keep the request sane. */
    private const val QUALITY = 80

    class EncodingException(message: String) : Exception(message)

    /**
     * Longest-edge target for a send of [imageCount] images. Text in a phone
     * screenshot stays legible at 900px, which is what makes the top of the
     * ladder affordable.
     */
    fun targetDimension(imageCount: Int): Int = when {
        imageCount <= 2 -> MAX_DIMENSION
        imageCount <= 5 -> 1100
        else -> 900
    }

    /**
     * Reads [uri], downscales it, and returns a `data:image/jpeg;base64,…` URL.
     *
     * @throws EncodingException when the image can't be read or encoded — the
     *   caller shows that message rather than blaming the picture's content.
     */
    fun dataUrl(context: Context, uri: Uri, maxDimension: Int = MAX_DIMENSION): String {
        val bitmap = decodeDownsampled(context, uri, maxDimension)
            ?: throw EncodingException("That image couldn't be read. Try a different screenshot.")

        val scaled = downscale(bitmap, maxDimension)

        val bytes = ByteArrayOutputStream().use { stream ->
            if (!scaled.compress(Bitmap.CompressFormat.JPEG, QUALITY, stream)) {
                throw EncodingException("That image couldn't be read. Try a different screenshot.")
            }
            stream.toByteArray()
        }

        val base64 = Base64.encodeToString(bytes, Base64.NO_WRAP)
        return "data:image/jpeg;base64,$base64"
    }

    /**
     * Decodes with `inSampleSize` so a 12MP photo never lands in memory at full
     * size just to be shrunk afterwards.
     */
    private fun decodeDownsampled(context: Context, uri: Uri, maxDimension: Int): Bitmap? {
        val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        context.contentResolver.openInputStream(uri)?.use {
            BitmapFactory.decodeStream(it, null, bounds)
        } ?: return null

        val longest = max(bounds.outWidth, bounds.outHeight)
        if (longest <= 0) return null

        var sample = 1
        while (longest / sample > maxDimension * 2) sample *= 2

        val options = BitmapFactory.Options().apply { inSampleSize = sample }
        return context.contentResolver.openInputStream(uri)?.use {
            BitmapFactory.decodeStream(it, null, options)
        }
    }

    /**
     * downscale returns the original when no scaling is needed, so it can't be
     * recycled here.
     */
    private fun downscale(bitmap: Bitmap, maxDimension: Int): Bitmap {
        val longest = max(bitmap.width, bitmap.height)
        if (longest <= maxDimension) return bitmap

        val ratio = maxDimension.toDouble() / longest
        return Bitmap.createScaledBitmap(
            bitmap,
            (bitmap.width * ratio).roundToInt().coerceAtLeast(1),
            (bitmap.height * ratio).roundToInt().coerceAtLeast(1),
            true
        )
    }
}
